### PR TITLE
feat: improved retry handling

### LINKED ISSUE #155: Improve the retry strategy
### Checklist

- [x] I agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md).

### Describe the problem you'd like to have solved

[Retry-After](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) is a standard header used by APIs to indicate when the SDK can retry.

The SDKs should:
* Honor this header on 429s
* Expose this header value in the error when received
* Fallback to exponential retry when this header is not available (e.g. not sent by the server on 429s or on e.g. 500s)
* Drop support for retrying based on `X-Rate-Limit-Reset` (currently only .NET SDK supports that), though still expose it in the logs

### Current State

|     |     |     |     |     |
| --- | --- | --- | --- | --- |
| **SDK** | **Retries on** | **Default Num Retries** | **Max Num Retries** | **State** |
| Java | 429s, 500s except for 501 | 3   | 15  | Does not consider headers. Does not implement exponential backoff. Adds a [fixed 100ms delay](https://github.com/openfga/java-sdk/blob/b7a0078ff4da51068111a3c2ef3d5ba0d909d064/src/main/java/dev/openfga/sdk/api/client/HttpRequestAttempt.java#L155-L159) to each retry |

#### Retry On
- For all requests:
  - Retry on 429s, honor `Retry-After` when sent by the server, falling back to exponential backoff when it's not.
  - Retry on 5xxs (except 501 not implemented), honor `Retry-After` when sent by the server, falling back to exponential backoff when it's not
  - Retry on network errors using exponential backoff

#### Max Allowable Retries

15

#### Default Number of Retries

SDKs: 3

#### Retry Parameters
1.  If `Retry-After` header is found, use it
    1.  if it is an integer, treat it as the number of seconds from now to retry, if it is &lt;1 from now or &gt;1800 from now (aka >30 min) - assume it is invalid and continue
    2.  if it is a date, parse it but if it is &lt;1 from now or &gt;1800 from now (aka >30 min) - assume it is invalid and continue
        
2.  If neither header is found, use exponential backoff but we'll add some jitter, so the retry is a random number between
    1.  `2^loopCount * 100ms and 2^(loopCount + 1) * 100ms`
    2.  if the result of (a) is > 120s, cap it at 120s which should happen between the 8th and 9th retry

That means:
* if retry-after header was returned and is valid, we’ll use it - so if it says in 4 min all good
* if retry-after header was not returned, we will retry at:
  * 100ms
  * 200ms
  * 400ms
  * 800ms
  * 1.6s
  * 3.2s
  * 6.4s
  * 12.8s
  * 25.6s
  * 51.2s
  * 102.4s
  * 120s ← at this point is is >4min since initial call
  * 120s
  * 120s
  * 120s

### Alternatives and current workarounds

_No response_

### References

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jimmyjames**
# Retry-After header support with enhanced retry strategy

## Summary

This PR implements comprehensive RFC 9110 compliant retry behavior with support for the `Retry-After` header, enhanced exponential backoff with jitter, network error recovery, and improved configuration validation with strict input validation.

Fixes #155

### 🔄 RFC 9110 Compliant Retry-After Header Support
- Parses both integer seconds and HTTP-date formats (RFC 1123)  
- Validates retry delays between 1-1800 seconds (30 minutes maximum)
- Prioritizes server-specified delays over exponential backoff
- Gracefully handles invalid header formats by falling back to exponential backoff

### 📈 Enhanced Exponential Backoff with Jitter
- Formula: `2^retryCount * minimumRetryDelay` with random jitter
- Prevents thundering herd problems
- Maximum delay capped at 120 seconds
- Default minimum retry delay: 100ms

### 🛡️ Unified Retry Strategy
- **Rate limiting (429)**: Always retried regardless of HTTP method
- **Server errors (5xx)**: Retried for ALL operations (except 501 Not Implemented)
  - Honors `Retry-After` header when present
  - Falls back to exponential backoff when `Retry-After` not available
- **Network errors**: Comprehensive retry support for connection timeouts, DNS failures, and network connectivity issues

### 🔧 Configuration Enhancements
- Maximum allowable retries enforced at 15 (with validation)
- Default retries remain 3 for backward compatibility
- Strict input validation for `minimumRetryDelay` configuration
- `Retry-After` header value exposed in `FgaError` objects

### 1. Strict Configuration Validation
**Previous**: `minimumRetryDelay` accepted null values and negative durations  
**New**: Throws `IllegalArgumentException` for null or negative values

```java
// Before: These would work (with runtime fallbacks)
var config = new Configuration().minimumRetryDelay(null);
var config = new Configuration().minimumRetryDelay(Duration.ofMillis(-100));

// After: These throw IllegalArgumentException  
var config = new Configuration().minimumRetryDelay(null);        // ❌ IllegalArgumentException
var config = new Configuration().minimumRetryDelay(Duration.ofMillis(-100)); // ❌ IllegalArgumentException

// Correct usage:
var config = new Configuration().minimumRetryDelay(Duration.ofMillis(100)); // ✅ Valid
```

### 2. Maximum Retry Validation
**Previous**: No validation on `maxRetries` value  
**New**: Throws `IllegalArgumentException` if `maxRetries > 15`

```java
// Before: This would work (dangerously)
var config = new Configuration().maxRetries(1000);

// After: This throws IllegalArgumentException
var config = new Configuration().maxRetries(1000); // ❌ IllegalArgumentException

// Correct usage:
var config = new Configuration().maxRetries(15); // ✅ Maximum allowed
```

### 3. Enhanced Error Information
**New**: `FgaError` now exposes `Retry-After` header value

```java
try {
    var response = client.check(request).get();
} catch (ExecutionException e) {
    if (e.getCause() instanceof FgaError) {
        FgaError error = (FgaError) e.getCause();
        String retryAfter = error.getRetryAfterHeader(); // New method
        if (retryAfter != null) {
            System.out.println("Server requested retry after: " + retryAfter);
        }
    }
}
```

### Basic Configuration
```java
var config = new ClientConfiguration()
    .maxRetries(3)                           // Default: 3, Maximum: 15
    .minimumRetryDelay(Duration.ofMillis(100)); // Default: 100ms, must be non-null and positive

var client = new OpenFgaClient(config);
```

#### Scenario 1: Rate Limiting (429)
```java
// Server responds with: 429 Too Many Requests, Retry-After: 5
// SDK behavior: Waits 5 seconds, then retries (up to 3 times)
client.check(request).get();
```

#### Scenario 2: Server Error with Retry-After Header
```java
// Server responds with: 503 Service Unavailable, Retry-After: 30
// SDK behavior: Waits 30 seconds, then retries
client.writeAuthorizationModel(model).get();
```

#### Scenario 3: Server Error without Retry-After Header
```java
// Server responds with: 500 Internal Server Error (no Retry-After header)
// SDK behavior: Uses exponential backoff (100ms, 200ms, 400ms) and retries
client.readAuthorizationModels().get();
```

#### Scenario 4: Network Error Recovery
```java
// Network connectivity issues (connection timeout, DNS failure, etc.)
// SDK behavior: Uses exponential backoff with jitter, retries up to configured maximum
client.listStores().get();
```

#### Scenario 5: Exponential Backoff Sequence
```java
// Without Retry-After header, delays follow: 100ms → 200ms → 400ms → 800ms → 1600ms → 3200ms...
// With jitter, actual delays vary between base and 2x base delay
// Maximum delay capped at 120 seconds
client.listObjects(request).get();
```

### New Utility Classes
- `RetryAfterHeaderParser`: RFC 9110 compliant header parsing for both seconds and HTTP-date formats
- `ExponentialBackoff`: Exponential backoff calculation with jitter and maximum delay cap
- `RetryStrategy`: Centralized retry decision logic for different error types

### Enhanced Classes
- `HttpRequestAttempt`: Updated retry logic with network error handling and proper minimum delay enforcement
- `Configuration`: Added strict validation for `minimumRetryDelay` and `maxRetries` parameters
- `FgaError`: Exposes `Retry-After` header value for improved observability

### Comprehensive Test Coverage
- 18+ retry configuration tests covering global and per-request scenarios
- Unit tests for all retry utilities (RetryAfterHeaderParser, ExponentialBackoff, RetryStrategy)
- Integration tests with mock `Retry-After` responses and network error simulation
- Edge case validation and error handling tests
- RFC 9110 compliance verification

## Migration Guide

1. **Update configuration validation**: Ensure `minimumRetryDelay` values are non-null and positive
   ```java
   // Replace this:
   config.minimumRetryDelay(null);
   
   // With this:
   config.minimumRetryDelay(Duration.ofMillis(100)); // or your preferred positive value
   ```

2. **Validate retry configuration**: Ensure `maxRetries` values are ≤ 15
   ```java
   // Replace this:
   config.maxRetries(50);
   
   // With this:
   config.maxRetries(15); // maximum allowed value
   ```

3. **Update error handling** (optional): Use new `getRetryAfterHeader()` method for custom retry logic
   ```java
   if (error instanceof FgaError) {
       String retryAfter = ((FgaError) error).getRetryAfterHeader();
       // Handle server-specified retry delays
   }
   ```

## Backward Compatibility

- Default retry count remains 3 (no change for existing code)
- All HTTP operations now benefit from unified retry behavior
- All existing configuration options preserved (with enhanced validation)
- No changes to public API surface except new `getRetryAfterHeader()` method
- Existing code using valid configuration values continues to work unchanged

## Performance and Reliability Improvements

- **Network resilience**: Automatic recovery from temporary network issues
- **Server load reduction**: Honors server-specified retry delays via `Retry-After` header
- **Thundering herd prevention**: Jitter in exponential backoff prevents synchronized retries
- **Fail-fast validation**: Configuration errors caught at setup time, not runtime
- **Comprehensive error handling**: Better visibility into retry behavior and server responses

## Testing

All existing tests pass (348+ tests) plus comprehensive new test coverage:

### New Test Coverage
- **Configuration validation tests**: Verify strict validation for `minimumRetryDelay` and `maxRetries`
- **Retry strategy tests**: Comprehensive coverage of retry decision logic
- **Network error tests**: Validation of retry behavior during network connectivity issues
- **Retry-After header tests**: RFC 9110 compliance for both seconds and HTTP-date formats
- **Integration tests**: End-to-end retry behavior with real HTTP mocking

### Test Categories
- Unit tests for retry utilities (100% coverage)
- Integration tests for retry configuration scenarios
- Edge case validation and error condition testing
- Performance tests for exponential backoff timing
- RFC 9110 compliance verification

## References

- [GitHub Issue #155](https://github.com/openfga/java-sdk/issues/155)
- [RFC 9110 Retry-After specification](https://www.rfc-editor.org/rfc/rfc9110.html#name-retry-after)
- [OpenFGA Java SDK Documentation](https://github.com/openfga/java-sdk)

### CODE REVIEW

**Review by evansims — APPROVED** [review]

**Review by rhamzeh — COMMENTED** [review]
Just a small change (updated the issue).

We later decided to retry on 5xx on all requests regardless of whether they change state or not.

Can we:
* For all (429s + 5xxs) except 501:
  * honor Retry-After when sent by the server, falling back to exponential backoff when it's not.
* For network errors (e.g. no internet connectivity/LB closes the connection/server disconnects) [these don't necessarily have an http reponse and status as the server didn't reply)
  * Retry using exponential backoff

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,28 @@
...
+- feat: Enhanced retry strategy with differentiated behavior for state-affecting operations
+- feat: Retry-After header value exposed in error objects for better observability
+
+### Changed
+- **BREAKING**: Maximum allowable retry count is now enforced at 15 (default remains 3)
+- **BREAKING**: State-affecting operations (POST, PUT, PATCH, DELETE) now only retry on 5xx errors when Retry-After header is present
+- **BREAKING**: FgaError now exposes Retry-After header value via getRetryAfterHeader() method
+
+### Technical Details
+- Implements RFC 9110 compliant Retry-After header parsing (supports both integer seconds and HTTP-date formats)
```
Comment (dyeam0) [inline CHANGELOG.md:16]: Oh, we can support `HTTP-date` format? That's cool. I thought during Resiliency Week discussions with @rhamzeh that there was difficulty in doing that.

**Review by dyeam0 — APPROVED** [review]
Docs are excellent. Looks ~~good~~ great!

**Comment by jimmyjames** [comment]
> Lovely PR + tests + documentation! ❤️
> 
> Just a small change (updated the issue).
> 
> We later decided to retry on 5xx on all requests regardless of whether they change state or not.
> 
> Can we:
> 
> * For all (429s + 5xxs) except 501:
>   
>   * honor Retry-After when sent by the server, falling back to exponential backoff when it's not.
> * For network errors (e.g. no internet connectivity/LB closes the connection/server disconnects) [these don't necessarily have an http reponse and status as the server didn't reply)
>   
>   * Retry using exponential backoff

Updated to remove state-specific handling and retry on network issues.

Also tested some basic and retry functionality using a test app created (not included) to exercise that code, worked as expected.

File: CHANGELOG.md (hunk 1 above)
Comment (rhamzeh) [reply CHANGELOG.md:15]: 🤔 Was there? This is how we built it for the Go SDKs.

The only problem was a personal one, date based retries are bad b/c you can never be sure the client and server are on the same clock and don't have any drift, but we support that in Go https://github.com/openfga/go-sdk/blob/main/internal/utils/retryutils/retryutils.go#L60-L63

Comment (dyeam0) [reply CHANGELOG.md:15]: Yes, personally I feel the same. Without tight control and a common NTP server, there are always drift issues. But I have no issues with supporting it, though. I must have been misremembering our conversation about `HTTP-date`.

**Review by dyeam0 — DISMISSED** [review]

File: src/test/java/dev/openfga/sdk/api/OpenFgaApiTest.java [hunk 2]
```diff
@@ -1913,7 +1921,8 @@ DEFAULT_STORE_ID, DEFAULT_AUTH_MODEL_ID, new WriteAssertionsRequest())
                 .get());
 
         // Then
-        mockHttpClient.verify().put(putUrl).called(1 + DEFAULT_MAX_RETRIES);
+        // Simplified logic: PUT requests now retry on 5xx errors (1 initial + 3 retries = 4 total)
+        mockHttpClient.verify().put(putUrl).called(4);
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/test/java/dev/openfga/sdk/api/OpenFgaApiTest.java:1925]: > Would it be possible to add a test w/ a network error here?

The http mocking lib makes this hard to do here, since it wraps excdeptions with an IllegalStateException. I think the tests we have in `HttpRequestAttemptRetryTest` in wiremock, including covers that scenario adequately and better simulates network issues

> And if I can ask, can we have a client request where the token exchange request fails once, to make sure all is recovering properly?

We have `exchangeOAuth2TokenWithRetriesSuccess` and `exchangeOAuth2TokenWithRetriesFailure` in `OAuth2ClientTest` - are those missing a scenario that should be tested for?

File: src/main/java/dev/openfga/sdk/util/RetryStrategy.java [hunk 3]
```diff
@@ -19,52 +19,41 @@
...
      * - All other status codes: Do not retry
      *
-     * @param request The HTTP request
+     * @param request The HTTP request (kept for API compatibility)
      * @param statusCode The HTTP response status code
-     * @param hasRetryAfterHeader Whether the response contains a valid Retry-After header
+     * @param hasRetryAfterHeader Whether the response contains a valid Retry-After header (kept for API compatibility)
      * @return true if the request should be retried, false otherwise
      */
     public static boolean shouldRetry(HttpRequest request, int statusCode, boolean hasRetryAfterHeader) {
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/dev/openfga/sdk/util/RetryStrategy.java]: Excellent catch - this appears to be a scenario missed in this PR (and not something we support yet in go or python?). I'll need to revisit the changes here to support the per-request configuration along with the global configuration and default behavior.

[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/dev/openfga/sdk/util/RetryStrategy.java]: I've updated the PR to honor the min retry scenario and add tests here: https://github.com/openfga/java-sdk/pull/186/commits/9cfce5222281e79af6b4d64b74bc8d5ff47f8759

File: src/main/java/dev/openfga/sdk/util/RetryStrategy.java [hunk 4]
```diff
@@ -60,7 +61,7 @@ public static boolean shouldRetry(int statusCode) {
      *
      * @param retryAfterDelay Optional delay from Retry-After header
      * @param retryCount Current retry attempt (0-based)
-     * @param minimumRetryDelay Minimum delay to enforce (can be null)
+     * @param minimumRetryDelay Minimum delay to enforce (also used as base delay for exponential backoff)
```
Comment (rhamzeh) [inline src/main/java/dev/openfga/sdk/util/RetryStrategy.java:64]:
```suggestion
     * @param minimumRetryDelay Base delay for exponential backoff
```

File: src/main/java/dev/openfga/sdk/api/configuration/Configuration.java [hunk 5]
```diff
@@ -37,9 +37,14 @@ public class Configuration implements BaseConfiguration {
     private static final Duration DEFAULT_READ_TIMEOUT = Duration.ofSeconds(10);
     private static final Duration DEFAULT_CONNECT_TIMEOUT = Duration.ofSeconds(10);
     private static final int DEFAULT_MAX_RETRIES = 3;
-    private static final Duration DEFAULT_MINIMUM_RETRY_DELAY = Duration.ofMillis(100);
     private static final int MAX_ALLOWABLE_RETRIES = 15;
 
+    /**
+     * Default minimum retry delay of 100ms.
+     * This value is also used as the default base delay for exponential backoff calculations.
```
Comment (rhamzeh) [inline src/main/java/dev/openfga/sdk/api/configuration/Configuration.java:44]:
```suggestion
     * This value is used as the default base delay for exponential backoff calculations.
```
